httpx/utils.py

```diff
@@ -212,6 +213,28 @@ def kv_format(**kwargs: typing.Any) -> str:
     return " ".join(f"{key}={value!r}" for key, value in kwargs.items())
 
 
+def should_not_be_proxied(url: "URL") -> bool:
+    """ Return True if url should not be proxied,
+    return False otherwise.
+    """
+    no_proxy = getproxies().get("no")
+    if not no_proxy:
+        return False
+    no_proxy_list = [host.strip() for host in no_proxy.split(",")]
+    for name in no_proxy_list:
+        if name == "*":
+            return True
+        if name:
+            name = name.lstrip(".")  # ignore leading dots
+            name = re.escape(name)
+            pattern = r"(.+\.)?%s$" % name
+            if re.match(pattern, url.host, re.I) or re.match(
+                pattern, url.authority, re.I
+            ):
```

Comment:
Just a note that we have uncompiled regexes here, which will harm performance as soon as a user users `NO_PROXY` (we'll keep compiling regex over and over on each request).

I can't think of an obvious way right now to turn this into a single compiled regex because the regex is composed from the `name` coming from `NO_PROXY`, so I won't turn this into a merge blocker. :)

Reply:
@florimondmanca FYI: `re.compile()` caches compiled patterns based on pattern+flags. :)

Reply:
Yes, but `re.match()` doesn't, or does it? 🤔

Reply:
[this](https://stackoverflow.com/a/452142) stackoverflow answer says the difference is negligible
[match](https://github.com/python/cpython/blob/7773d391f8261a97af502ecfa74a2a588ee05f36/Lib/re.py#L189) -> [_compile](https://github.com/python/cpython/blob/7773d391f8261a97af502ecfa74a2a588ee05f36/Lib/re.py#L287) code suggests patterns are cached
notice that due to the no_proxy host (name) being in the regex the same pattern is used twice, not very many times